### PR TITLE
Turn `allow-{newer,older}` into an accumulating field

### PR DESCRIPTION
**Pull request by hvr**
This fixes the inconvenience occuring in `cabal.project` files of
separate `allow-*` statements overwriting each other. Currently,
without this patch,

    allow-newer: base
    allow-newer: directory

is *NOT* equivalent to

    allow-newer: base, directory

but rather to the last occurence of `allow-newer`, i.e.

    allow-newer: directory

This behaviour has been present since cabal 1.24

----

Please include the following checklist in your PR:

* [ ] Patches conform to the [coding conventions](https://github.com/haskell/cabal/#conventions).
* [ ] Any changes that could be relevant to users have been recorded in the changelog.
* [ ] The documentation has been updated, if necessary.

Please also shortly describe how you tested your change. Bonus points for added tests!

### CODE REVIEW

**Comment by hvr** [comment]
@23Skidoo This one would be a candidate for backporting to 2.0

**Comment by ezyang** [comment]
LGTM, but it would be nice if we had a test...

**Comment by 23Skidoo** [comment]
Hmm, I thought I've fixed this before, though maybe not for `cabal.project`.

**Comment by hvr** [comment]
@23Skidoo you're probably referring to #3165

**Comment by 23Skidoo** [comment]
Yep, just found that PR myself.

File: cabal-install/Distribution/Client/ProjectConfig/Legacy.hs [hunk 1]
```diff
@@ -1253,6 +1253,13 @@ listFieldWithSep separator name showF readF get' set =
     set' xs b = set (get' b ++ xs) b
     showF'    = separator . map showF
 
+monoidField :: Monoid a => String -> (a -> Doc) -> ReadP a a
```
Comment (23Skidoo) [inline cabal-install/Distribution/Client/ProjectConfig/Legacy.hs]: Would be nice to have a Haddock comment here.

**Comment by hvr** [comment]
@ezyang jfyi, I have not forgotten about the tests; I plan to provide a couple lateron

**Pull request merged**